File: projects/dejavu.kt/dejavu-engine/src/test/kotlin/dejavu/engine/ConformanceTest.kt
package dejavu.engine

import dejavu.Dejavu
import dejavu.types.normalize
import kotlinx.serialization.json.*
import java.io.File
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class ConformanceTest {
    private val root: File = findConformanceRoot()

    @Test
    fun t1CasesParseAndRenderFromExpectedIr() {
        assertTrue(root.isDirectory, "missing $root")
        val json = Json { ignoreUnknownKeys = true }
        root.listFiles()!!.filter { it.isDirectory }.sortedBy { it.name }.forEach { dir ->
            val input = File(dir, "input.dejavu").readText()
            val expectedIrText = File(dir, "expected.ir.json").readText()
            val expectedIr = json.parseToJsonElement(expectedIrText).jsonObject
            val ctx = json.parseToJsonElement(File(dir, "context.ctx.json").readText()).jsonObject
            val expectedOut = File(dir, "expected.out.txt").readText()

            val got = Dejavu.parse(input)
            assertEquals(normalize(expectedIr), normalize(got), "IR mismatch in ${dir.name}")

            val out = Dejavu.render(expectedIr, ctx)
            assertEquals(expectedOut, out, "render mismatch in ${dir.name}")
        }
    }

    companion object {
        private val REL = "specifications/conformance/t1"

        private fun findConformanceRoot(): File {
            val markers = listOf(
                File("../../../$REL"),
                File("../../../../$REL"),
                File(REL),
            )
            for (c in markers) {
                val f = c.canonicalFile
                if (f.isDirectory) return f
            }
            var dir = File(System.getProperty("user.dir")).canonicalFile
            repeat(10) {
                val candidate = File(dir, REL)
                if (candidate.isDirectory) return candidate
                dir = dir.parentFile ?: return@repeat
            }
            return File("../../../$REL").canonicalFile
        }
    }
}

File: projects/dejavu.ts/dejavu-engine/tests/conformance.test.ts
import { describe, it, expect } from "vitest";
import { readFileSync, readdirSync } from "node:fs";
import { join, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import { compareIrJson, engine, normalizeIrJson, renderIr, type IrDocument } from "../src";

const root = join(
    dirname(fileURLToPath(import.meta.url)),
    "../../../../specifications/conformance/t1",
);

describe("specifications/conformance/t1", () => {
    const cases = readdirSync(root, { withFileTypes: true })
        .filter((d) => d.isDirectory())
        .map((d) => d.name)
        .sort();

    for (const name of cases) {
        it(name, () => {
            const dir = join(root, name);
            const input = readFileSync(join(dir, "input.dejavu"), "utf8");
            const expectedIr = readFileSync(join(dir, "expected.ir.json"), "utf8");
            const ctx = JSON.parse(readFileSync(join(dir, "context.ctx.json"), "utf8"));
            const expectedOut = readFileSync(join(dir, "expected.out.txt"), "utf8");

            const doc = engine.parse(input);
            expect(
                compareIrJson(JSON.stringify(doc), expectedIr),
                `IR mismatch\n${JSON.stringify(normalizeIrJson(JSON.stringify(doc)), null, 2)}`,
            ).toBe(true);

            const out = renderIr(JSON.parse(expectedIr) as IrDocument, ctx);
            expect(out).toBe(expectedOut);
        });
    }
});
